File: functional/ocaml/pow_pb_simpl.ml
let rec pow' (x,n) = 
  match n with
  | 0 -> (1, fun s -> (0,0))
  | _ -> 
    let (y, pb) = pow' (x, n-1) in
    (x*y,
      fun s ->
        let (x', n') = pb (x*s) in
        (x'+y*s, n')
    )

let _ = 
  let (z, z_pb) = pow' (4,3) in
  let (xt, nt) = z_pb 1 in
  Printf.printf "z = %d, x' = %d, n' = %d\n" z xt nt

File: eval/functional/haskell/pow_pb_simpl.hs
pow' (x, n) = 
    if n == 0 then
        (1, \s -> (0,0))
    else
        let (y,y_pb) = pow' (x, n-1) in
        (x*y,
            \s ->
                let (a1,a2) = (y_pb (x*s)) in
                    (a1+y*s,a2)
        )
        

main = 
    let (z, z_pb) = pow' (400,30000) in
    do print z
       print (z_pb 1)
